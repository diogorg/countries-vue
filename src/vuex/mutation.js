import {getCountriesFromCache, getCountriesFromApi} from './model/get-countries'
import {apiFailNotification, apiSuccessNotification} from '@/notifications'
import {filterCountries} from './model/filter-countries'

export const mutation = {
    selectCountry(state, { country }) {
        state.dialog = true
        state.selected = country
    },
    closeDialog(state) {
        state.dialog = false
        state.selected = null
    },
    filterCountries(state, { name, region, lang }) {
        state.countries = filterCountries(state.total, name, region, lang)
    },
    getAllCountriesAndRegions(state, { quasar }) {
        const data = getCountriesFromCache() || getCountriesFromApi()
        if (!data) {
            apiFailNotification(quasar)
            return
        }
        apiSuccessNotification(quasar)
        state.countries = data.countries
        state.total = data.total
        state.regions = data.regions
        state.langs = data.langs
        return
    }
}