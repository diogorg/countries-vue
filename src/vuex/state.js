export const state = {
    total: [],
    countries: [],
    regions: ['ALL'],
    langs: ['ALL'],
    dialog: false,
    selected: null
}