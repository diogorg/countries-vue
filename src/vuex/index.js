import Vuex from 'vuex'
import {state} from './state'
import {mutation} from './mutation'

export const store = new Vuex.Store({
    state: () => state,
    mutations: mutation,
    actions: {}
})