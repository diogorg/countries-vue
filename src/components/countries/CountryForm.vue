<template>
    <div class="filter-col q-pa-sm">
        <form v-on:keyup.enter="filterCountries" class="filter-form q-pa-sm q-pb-md shadow-1">
            <h3 class="q-pa-sm">Filter Countries</h3>
            <q-input outlined v-model="name" label="Country Name" class="q-ma-sm" />
            <q-select outlined v-model="region" :options="regions" label="Region" class="q-ma-sm" />
            <q-select class="q-ma-sm" outlined v-model="lang" use-input hide-selected fill-input input-debounce="0"
                :options="options" @filter="filterLangs" label="Languages" emit-value map-options>
                <template v-slot:no-option>
                    <q-item>
                        <q-item-section class="text-grey">
                            No results
                        </q-item-section>
                    </q-item>
                </template>
            </q-select>
            <q-btn @click="filterCountries" class="glossy q-ma-sm" color="deep-orange" icon="search" label="Filter" />
        </form>
    </div>
</template>

<script>
import { mapState } from 'vuex'
export default {
    name: 'CountryForm',
    data: () => ({
        name: '',
        region: null,
        lang: null,
        options: []
    }),
    mounted() {
        this.options = this.langs
    },
    methods: {
        filterLangs(val, update) {
            update(() => {
                if (!val) {
                    this.options = this.langs
                    return
                }
                const needle = val.toLowerCase()
                this.options = this.langs.filter(v => v.toLowerCase().indexOf(needle) > -1)
            })
        },
        async filterCountries() {
            this.$q.loading.show({
                message: 'Filtering the countries',
                boxClass: 'bg-grey-2 text-grey-9',
                spinnerColor: 'primary'
            })
            setTimeout(() => {
                this.$store.commit('filterCountries', { name: this.name, region: this.region, lang: this.lang })
                this.$q.loading.hide()
            }, 700)
        }
    },
    computed: {
        ...mapState(['regions', 'langs']),
    },
    components: {},
    setup() { }
}
</script>

<style scoped>
.filter-col {
    width: 300px;
    position: relative;
}

.filter-form {
    background: #fff;
    width: fit-content;
    min-height: 100px;
    border: 1px dashed #bbb;
    text-align: center;
    position: fixed;
}

.filter-form h3 {
    text-align: center;
    width: 100%;
    font-size: 21px;
    margin: 0px;
}
</style>