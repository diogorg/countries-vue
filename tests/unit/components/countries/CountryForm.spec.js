import { shallowMount, mount } from '@vue/test-utils'
import CountryForm from '@/components/countries/CountryForm.vue'
import { createStore } from 'vuex'
import { components } from './basic-components'

const state = {
    total: [],
    countries: [],
    regions: ['ALL', 'R1', 'R2', 'R3'],
    langs: ['ALL', 'L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7'],
    dialog: false,
    selected: null
}
const mutations = {
    filterCountries: jest.fn()
}
const store = createStore({
    state() {
        return state
    },
    mutations
})

const quasar = {
    loading: {
        show: jest.fn(),
        hide: jest.fn(),
    }
}

const global = {
    components,
    mocks: {
        $store: store,
        $q: quasar
    }
}

describe('CountryForm.vue', () => {
    it('Should match snapshot', () => {
        const wrap = mount(CountryForm, {
            global
        })
        expect(wrap.html()).toMatchSnapshot()
    })

    it('Should Filter After Click on Button', async () => {
        const spy = jest.spyOn(store, 'commit')
        const spyMutation = jest.spyOn(mutations, 'filterCountries')
        const wrap = shallowMount(CountryForm, {
            global
        })
        const button = wrap.find('#filter-button')
        await button.trigger('click')

        expect(spy).toHaveBeenCalledTimes(1)
        expect(spyMutation).toHaveBeenCalledTimes(1)
        expect(quasar.loading.show).toHaveBeenCalledTimes(1)
        expect(quasar.loading.show).toBeCalledWith({
            message: 'Filtering the countries',
            boxClass: 'bg-grey-2 text-grey-9',
            spinnerColor: 'primary'
        })
        expect(quasar.loading.hide).toHaveBeenCalledTimes(1)
    })

    it('Should Filter Langs Options', async () => {
        const wrap = await mount(CountryForm, {
            global
        })

        expect(wrap.vm.options).toEqual(['ALL', 'L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7'])
        await wrap.vm.filterLangs('L1', jest.fn())
        expect(wrap.vm.options).toEqual(['L1'])
    })

    it('Should Keep Langs Options When Filter is Null', async () => {
        const wrap = await mount(CountryForm, {
            global
        })

        expect(wrap.vm.options).toEqual(['ALL', 'L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7'])
        await wrap.vm.filterLangs(null, jest.fn())
        expect(wrap.vm.options).toEqual(['ALL', 'L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7'])
    })

    it('Should have H3', async () => {
        const wrap = await mount(CountryForm, {
            global
        })
        expect(wrap.html()).toContain('<h3 class="q-pa-sm">Filter Countries</h3>')
    })
})