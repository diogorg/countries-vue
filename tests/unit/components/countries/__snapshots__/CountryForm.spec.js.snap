// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryForm.vue Should match snapshot 1`] = `
<div class="filter-col q-pa-sm">
  <form class="filter-form q-pa-sm q-pb-md shadow-1">
    <h3 class="q-pa-sm">Filter Countries</h3>
    <q-input outlined="" modelvalue="" label="Country Name" class="q-ma-sm"></q-input>
    <q-select outlined="" options="ALL,R1,R2,R3" label="Region" class="q-ma-sm"></q-select>
    <q-select id="filter-lang-select" class="q-ma-sm" outlined="" use-input="" hide-selected="" fill-input="" input-debounce="0" options="" label="Languages" emit-value="" map-options=""></q-select><button id="filter-button" class="glossy q-ma-sm" color="deep-orange" icon="search" label="Filter"></button>
  </form>
</div>
`;
